File: src/main/java/com/bazel-diff/BazelTarget.java
import com.google.devtools.build.lib.query2.proto.proto2api.Build;

interface BazelTarget {
    boolean hasRule();
    BazelRule getRule();
    boolean hasSourceFile();
    String getSourceFileName();
}

class BazelTargetImpl implements BazelTarget {
    private Build.Target target;

    public BazelTargetImpl(Build.Target target) {
        this.target = target;
    }

    @Override
    public boolean hasRule() {
        return target.hasRule();
    }

    @Override
    public BazelRule getRule() {
        if (this.hasRule()) {
            return new BazelRuleImpl(target.getRule());
        }
        return null;
    }

    @Override
    public boolean hasSourceFile() {
        return target.hasSourceFile();
    }

    @Override
    public String getSourceFileName() {
        if (this.hasSourceFile()) {
            return this.target.getSourceFile().getName();
        }
        return null;
    }
}

File: cli/src/main/kotlin/com/bazel_diff/bazel/BazelTarget.kt
package com.bazel_diff.bazel

import com.google.devtools.build.lib.query2.proto.proto2api.Build

sealed class BazelTarget(private val target: Build.Target) {
    class SourceFile(target: Build.Target) : BazelTarget(target) {
        init {
            assert(target.hasSourceFile())
        }

        val sourceFileName: String = target.sourceFile.name
        override val name: String
            get() = sourceFileName
    }

    class Rule(target: Build.Target) : BazelTarget(target) {
        init {
            assert(target.hasRule())
        }

        val rule: BazelRule = BazelRule(target.rule)
        override val name: String
            get() = rule.name
    }

    class GeneratedFile(target: Build.Target) : BazelTarget(target) {
        init {
            assert(target.hasGeneratedFile())
        }

        val generatedFileName: String = target.generatedFile.name
        val generatingRuleName: String = target.generatedFile.generatingRule
        override val name: String
            get() = generatedFileName
    }

    val type: BazelTargetType
        get() = when (target.type) {
            Build.Target.Discriminator.RULE -> BazelTargetType.RULE
            Build.Target.Discriminator.SOURCE_FILE -> BazelTargetType.SOURCE_FILE
            Build.Target.Discriminator.GENERATED_FILE -> BazelTargetType.GENERATED_FILE
            Build.Target.Discriminator.PACKAGE_GROUP -> BazelTargetType.PACKAGE_GROUP
            Build.Target.Discriminator.ENVIRONMENT_GROUP -> BazelTargetType.ENVIRONMENT_GROUP
            else -> BazelTargetType.UNKNOWN
        }

    abstract val name: String
}

